### DISCUSSION TITLE
Indicate new block with indentation (as alternative to newline)

### DISCUSSION THREAD

**Discussion by hellux** [jgm/djot#210] [Ideas]
This might have been discussed before as it is a quite simple idea, but it
could potentially relieve a requirement that seems to be an annoyance 
to many: newline before lists. Previous discussions: #137, #161, #170, #173.

When hardwrapping text, one does usually not add additional indentation to
lines that continue a paragraph, so an additional indent could be used to
indicate that it is the start of a new block, e.g. a list item.

So

```
My favorite number is probably the number
1. It's the smallest natural number..
```

would be unchanged and form a single paragraph, but

```
My favorite two numbers are probably
 1. because it is #1,
 2. because it is #1 after 1.
```

would create a list. It seems fairly natural to have indentation here, as lists
that are immediately after a paragraph are usually a "subtopic" of that
paragraph.

A larger example with just compact lists of topics, in markdown:

```
**topic 1**:
- some text wrapping on
  next line
  - aa
- b
  - ba
  - bb
- c

**topic 2**:
- [ ] a
  - [x] aa
- b
  1. b1
    a) b1a
    b) b1b
  2. bb
- c
```

same example, with proposed syntax

```
*topic 1*:
  - some text wrapping on
    next line
     - aa
  - b
     - ba
     - bb
  - c

**topic 2**:
  - [ ] a
     - [x] aa
  - b
     1. b1
         a) b1a
         b) b1b
     2. bb
  - c
```

One could discuss where the indentation starts for each type of list here.

In the spirit of uniformity, this idea could perhaps be generalized to start any
type of block and not just lists, for example, it could allow e.g
````
para para
 ```
 immediate code block
 ```

para para
 |immediate|table|
````

It could also generalize to break other blocks than paragraphs, but the
current implementation does not seem to mind breaking non-paragraphs
without newlines already? E.g. headings:
```
# heading
- this is a list, not a continuation of the heading text
```

**Comment by matklad** [comment]
> would create a list. It seems fairly natural to have indentation here, as lists
that are immediately after a paragraph are usually a "subtopic" of that
paragraph.

To rephrase this, this is inline with “source of the document looks roughly like the rendered document”. When lists are rendered, they are typically indented relative to the paragraphs (the marker itself is indented).

**Reply by hellux** [reply]
That's true, I guess most formats/renderers (at least html, latex, man) do add padding to the left of list markers by default.

Personally, I sometimes naturally indent lists in markdown also, and then get confused when some implementations turn it into an indented code block.

**Comment by jgm** [comment]
If this were allowed, then I think people would assume you could do
```
This is a list
  1. one
  2. two
and here we are outside the list.
```
Then they'd be confused when, because of "laziness", that last line is added to the second list item.

I've been tempted to use indentation for another purpose: distinguishing block quotes that are part of a paragraph from those that are separate blocks.  (If the block quote is part of a paragraph, then the following content should not be indented in formatted output, but if it not, the following content should be indented. In my own markdown writing with pandoc I end up putting raw tex `\noindent` all over the place to avoid unwanted paragraph indents after embedded block quotes, and this is not good.)  See @bpj's suggestion at #50.  I'm not sure the two proposals about indentation are compatible.

**Reply by hellux** [reply]
> If this were allowed, then I think people would assume you could do
> 
> ```
> This is a list
>   1. one
>   2. two
> and here we are outside the list.
> ```
> 
> Then they'd be confused when, because of "laziness", that last line is added to the second list item.

You could have that any indentation change could indicate a break. So going back in indentation would break the paragraph in the last list item in this case. To not break, the next line must start with the same indent, i.e.

```
This is a list
  1. one
  2. two with
     continuation
paragraph continue
```

Though, one can discuss what might happen if the indent is inbetween these or higher. Will an extra indented paragraph continue or create a new one?

> I've been tempted to use indentation for another purpose: distinguishing block quotes that are part of a paragraph from those that are separate blocks. (If the block quote is part of a paragraph, then the following content should not be indented in formatted output, but if it not, the following content should be indented. In my own markdown writing with pandoc I end up putting raw tex `\noindent` all over the place to avoid unwanted paragraph indents after embedded block quotes, and this is not good.) See @bpj's suggestion at #50. I'm not sure the two proposals about indentation are compatible.

I guess you could merge and generalize both of these proposals by adding any indented element under the current paragraph, so e.g.

```
abc
  > quote
def
  - one
  - two
ghi
   klm
nop
```
turns into
```
para
    str text="abc"
    block_quote
        para
            str="quote"
    str text="def"
    bullet_list
        list_item
            para
                str text="one"
        list_item
            para
                str text="two"
    para
        str text="klm"
    str tex="nop"
```

note: html does not seem to like placing blockquotes/lists/paragraphs inside paragraphs, the AST may be represented like this, but the rendered output may need some flattening depending on the format.

This kind of generalizes the existing nesting of lists with indentation for all blocks then.

**Comment by jgm** [comment]
> you could have that any indentation change could indicate a break.

This would be changing current "laziness" which allows this to be a list item

```
  - one list item
with continued content
```

**Reply by jgm** [reply]
> adding any indented element under the current paragraph

This would require making paragraphs into block containers (i.e. they could have block-level as well as inline-level children). I'm reluctant to do that because most output formats (e.g. HTML) don't allow block-level children under a paragraph.

**Reply by hellux** [reply]
> > you could have that any indentation change could indicate a break.
> 
> This would be changing current "laziness" which allows this to be a list item
> 
> ```
>   - one list item
> with continued content
> ```

Yeah, it would become quite strict. It would assume that any hard wrapping is "properly" indented.

Alternatively, one could just allow the first part, i.e. adding indent to indicate a break of the paragraph. It would then mostly work similar to Markdown: no newline needed to start a list, but a newline needed to end the list. So it might not seem too confusing for those who are familiar with Markdown at least.

> > adding any indented element under the current paragraph
> 
> This would require making paragraphs into block containers (i.e. they could have block-level as well as inline-level children). I'm reluctant to do that because most output formats (e.g. HTML) don't allow block-level children under a paragraph.

Yeah, I guess it would not map 1:1 with most formats in that case. It could only be used to change some part of the rendering, like omitting padding for blockquotes like you mentioned earlier.

To keep a close 1:1 mapping between the AST and output format, an alternative option would be to make the lists/blockquotes standalone outside the paragraph, like currently, but mark them as "inline" (or e.g. standalone=false) if they are indented relative to the paragraph (and thus disallow any further nesting). One could then e.g. choose to render "inline" blockquotes differently from standalone blockquotes.

For example:

```
inline quote:

  > newline before optional

inline list:
  - one
  - two

standalone quote:

> standalone quote, due to no indent

new paragraph
```
yields
```
doc
  para
    str text="inline quote:"
  block_quote inline=true
    para
      str text="newline before optional"
  para
    str text="inline list:"
  bullet_list inline=true tight=true style="-"
    list_item
      para
        str text="one"
    list_item
      para
        str text="two"
  para
    str text="standalone quote:"
  block_quote inline=false
    para
      str text="standalone quote, due to no indent"
  para
    str text="new paragraph"
```
